crates/bevy_utils/src/intern.rs

```diff
@@ -0,0 +1,135 @@
+    // you don't remove items (which we don't).
+    // This allows us to have O(~1) hashing and map each entry to a stable index.
+    RwLock<IndexSet<T>>,
+);
+
+/// The type returned from [`Interner::get`](Interner#method.get).
+///
+/// Will hold a lock on the interner until this guard gets dropped.
+pub type InternGuard<'a, L> = parking_lot::MappedRwLockReadGuard<'a, L>;
+
+impl<T: Clone + Hash + Eq> Interner<T> {
+    pub const fn new() -> Self {
+        Self(RwLock::new(IndexSet::with_hasher(FixedState)))
+    }
+
+    /// Interns a value, if it was not already interned in this set.
+    ///
+    /// Returns an integer used to refer to the value later on.
+    pub fn intern(&self, val: &T) -> usize {
+        use parking_lot::RwLockUpgradableReadGuard as Guard;
+
+        // Acquire an upgradeable read lock, since we might not have to do any writing.
+        let set = self.0.upgradable_read();
+
+        // If the value is already interned, return its index.
+        if let Some(idx) = set.get_index_of(val) {
+            return idx;
+        }
+
+        // Upgrade to a mutable lock.
+        let mut set = Guard::upgrade(set);
+        let (idx, _) = set.insert_full(val.clone());
+        idx
+    }
+
+    /// Gets a reference to the value with specified index.
+    pub fn get(&self, idx: usize) -> Option<InternGuard<T>> {
+        RwLockReadGuard::try_map(self.0.read(), |set| set.get_index(idx)).ok()
+    }
+}
+
+struct TypeMap(StableHashMap<TypeId, Box<dyn std::any::Any + Send + Sync>>);
+
+impl TypeMap {
+    pub const fn new() -> Self {
+        Self(StableHashMap::with_hasher(FixedState))
+    }
+
+    pub fn insert<T: Send + Sync + 'static>(&mut self, val: T) -> Option<impl Drop> {
+        self.0.insert(TypeId::of::<T>(), Box::new(val))
+    }
+    pub fn get<T: 'static>(&self) -> Option<&T> {
+        let val = self.0.get(&TypeId::of::<T>())?.as_ref();
+        // SAFETY: `val` was keyed with the TypeId of `T`, so we can cast it to `T`.
+        Some(unsafe { &*(val as *const _ as *const T) })
+    }
+    pub fn get_mut<T: 'static>(&mut self) -> Option<&mut T> {
+        let val = self.0.get_mut(&TypeId::of::<T>())?.as_mut();
+        // SAFETY: `val` was keyed with the TypeId of `T`, so we can cast it to `T`.
+        Some(unsafe { &mut *(val as *mut _ as *mut T) })
+    }
+}
+
+/// Data structure used to intern a set of values of any given type.
+///
+/// If you just need to store a single concrete type, [`Interner`] is more efficient.
+pub struct AnyInterner(
+    // This type-map stores instances of `IndexSet<T>`, for any `T`.
+    RwLock<TypeMap>,
+);
+
+impl AnyInterner {
+    pub const fn new() -> Self {
+        Self(RwLock::new(TypeMap::new()))
+    }
+
+    /// Interns a value, if it was not already interned in this set.
+    ///
+    /// Returns an integer used to refer to the value later on.
+    pub fn intern<L>(&self, val: &L) -> usize
+    where
+        L: Clone + Hash + Eq + Send + Sync + 'static,
```

Comment:
I don't understand why the `Send` and `Sync` bounds are needed here. I'm not sure they're wrong, but if they're necessary we should have a small comment explaining them.

Reply:
I think you're right, `Sync` at least should be unnecessary.